--- freetype-2.0.1/ttmkfdir2/ttf.cpp.foundry	Thu Jan 11 19:43:50 2001
+++ freetype-2.0.1/ttmkfdir2/ttf.cpp	Thu Jan 11 19:44:02 2001
@@ -282,7 +282,7 @@
 	{ "FSI ", "FSI Fonts und Software GmbH"},
 	{ "FTFT", "FontFont"},
 	{ "FWRE", "Fontware Limited"},
-	{ "GALA", "Galapagos Design Group, Inc."},
+	{ "GALA", "Galapagos"},
 	{ "GD"  , "GD Fonts"},
 	{ "GD  ", "GD Fonts"},
 	{ "GLYF", "Glyph Systems"},
